### PR TITLE
Allow null value in query string KVP per RFC 3986

### PR DESCRIPTION
**Pull request by robertmiles3**
Fixes https://github.com/aspnet/Home/issues/2679.

This not only fixes the issue specified, but also allows a null KVP value in other methods per [RFC 3986](https://tools.ietf.org/html/rfc3986#section-3.4) referenced by @jkotalik.

I added a test as well as enhanced existing tests to validate the null option.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Http.Abstractions/QueryString.cs [hunk 1]
```diff
@@ -159,15 +163,21 @@ public static QueryString Create(IEnumerable<KeyValuePair<string, StringValues>>
         {
             var builder = new StringBuilder();
             bool first = true;
+
             foreach (var pair in parameters)
             {
+                // If nothing in this pair.Values, append null value and continue
+                if (pair.Value.Count == 0)
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Http.Abstractions/QueryString.cs]: StringValues.IsNullOrEmpty

File: src/Microsoft.AspNetCore.Http.Abstractions/QueryString.cs [hunk 2]
```diff
@@ -209,7 +215,10 @@ public QueryString Add(string name, string value)
             builder.Append("&");
             builder.Append(UrlEncoder.Default.Encode(name));
             builder.Append("=");
-            builder.Append(UrlEncoder.Default.Encode(value));
```
Comment (campersau) [inline src/Microsoft.AspNetCore.Http.Abstractions/QueryString.cs]: You can also use `AppendKeyValuePair` with `first = false` here.

File: src/Microsoft.AspNetCore.Http.Abstractions/QueryString.cs (hunk 1 above)
Comment (robertmiles3) [reply src/Microsoft.AspNetCore.Http.Abstractions/QueryString.cs]: Good call. Committed.

File: src/Microsoft.AspNetCore.Http.Abstractions/QueryString.cs (hunk 2 above)
Comment (robertmiles3) [reply src/Microsoft.AspNetCore.Http.Abstractions/QueryString.cs]: Agreed. Committed.

File: src/Microsoft.AspNetCore.Http.Abstractions/QueryString.cs [hunk 3]
```diff
@@ -121,12 +121,19 @@ public static QueryString Create(string name, string value)
             {
                 throw new ArgumentNullException(nameof(name));
             }
-            if (value == null)
+
+            return new QueryString($"?{UrlEncoder.Default.Encode(name)}={(string.IsNullOrEmpty(value) ? null : UrlEncoder.Default.Encode(value))}");
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Http.Abstractions/QueryString.cs]: nit: Add some indentation here, bit hard to read.

File: src/Microsoft.AspNetCore.Http.Abstractions/QueryString.cs [hunk 4]
```diff
@@ -121,12 +121,19 @@ public static QueryString Create(string name, string value)
...
+        }
+
+        private static void AppendKeyValuePair(StringBuilder builder, string key, string value, bool first)
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Http.Abstractions/QueryString.cs]: Move the private method to the end of the file.

File: src/Microsoft.AspNetCore.Http.Abstractions/QueryString.cs [hunk 5]
```diff
@@ -195,21 +205,14 @@ public QueryString Add(string name, string value)
...
             {
                 return Create(name, value);
             }
 
             var builder = new StringBuilder(Value);
-            builder.Append("&");
-            builder.Append(UrlEncoder.Default.Encode(name));
-            builder.Append("=");
-            builder.Append(UrlEncoder.Default.Encode(value));
+            AppendKeyValuePair(builder, name, value, false);
```
Comment (khellang) [inline src/Microsoft.AspNetCore.Http.Abstractions/QueryString.cs]: nit: `first: false`?

Comment (robertmiles3) [reply src/Microsoft.AspNetCore.Http.Abstractions/QueryString.cs]: Agreed. Committed, thanks.

File: src/Microsoft.AspNetCore.Http.Abstractions/QueryString.cs (hunk 3 above)
Comment (robertmiles3) [reply src/Microsoft.AspNetCore.Http.Abstractions/QueryString.cs]: I added a few spaces for you. I'm not sure exactly what you had in mind, so if you want further readability tweaks just let me know.

File: src/Microsoft.AspNetCore.Http.Abstractions/QueryString.cs [hunk 6]
```diff
@@ -121,12 +121,8 @@ public static QueryString Create(string name, string value)
             {
                 throw new ArgumentNullException(nameof(name));
             }
-            if (value == null)
-            {
-                throw new ArgumentNullException(nameof(value));
-            }
 
-            return new QueryString($"?{UrlEncoder.Default.Encode(name)}={UrlEncoder.Default.Encode(value)}");
+            return new QueryString($"?{UrlEncoder.Default.Encode(name)}={( string.IsNullOrEmpty(value) ? null : UrlEncoder.Default.Encode(value) )}");
```
[the remark this answers is not in the document]
Comment (robertmiles3) [reply src/Microsoft.AspNetCore.Http.Abstractions/QueryString.cs]: Committed. How's that?

[the remark this answers is not in the document]
Comment (Tratcher) [reply src/Microsoft.AspNetCore.Http.Abstractions/QueryString.cs]: Better, thanks.

**Comment by davidfowl** [comment]
@robertmiles3 Thanks for the PR, we should really add some benchmarkdotnet microbenchmarks for this code.

**Comment by robertmiles3** [comment]
My first "real" PR in the books for dotnet; that's key. Thanks everyone for the feedback.

**Pull request merged**